includes/class-wc-payments-incentives-service.php

```diff
@@ -0,0 +1,208 @@
+				$menu[ $index ][0] .= WC_Payments_Admin::MENU_NOTIFICATION_BADGE; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+				break;
+			}
+		}
+	}
+
+	/**
+	 * Adds allowed promo notes from eligible incentive.
+	 *
+	 * @param array $promo_notes Current allowed promo notes.
+	 * @return array Updated allowed promo notes.
+	 */
+	public function allowed_promo_notes( $promo_notes = [] ): array {
+		if ( $this->get_cached_connect_incentive() ) {
+			/**
+			 * Suppress psalm error because we already check that `id` exists in `is_valid_cached_incentive`.
+			 *
+			 * @psalm-suppress PossiblyNullArrayAccess */
+			$promo_notes[] = $this->get_cached_connect_incentive()['id'];
+		}
+
+		return $promo_notes;
+	}
+
+	/**
+	 * Gets and caches eligible connect incentive from the server.
+	 *
+	 * @param bool $force_refresh Forces data to be fetched from the server, rather than using the cache.
+	 *
+	 * @return array|null List of incentives or null.
+	 */
+	public function get_cached_connect_incentive( bool $force_refresh = false ): ?array {
+		// Return early if there is an account connected.
+		if ( WC_Payments::get_account_service()->is_stripe_connected() ) {
+			return null;
+		}
+
+		// Return early if the country is not supported.
+		if ( ! array_key_exists( WC()->countries->get_base_country(), WC_Payments_Utils::supported_countries() ) ) {
+			return null;
+		}
+
+		$incentive = $this->database_cache->get_or_add(
+			Database_Cache::CONNECT_INCENTIVE_KEY,
+			[ $this, 'fetch_connect_incentive' ],
+			'is_array',
+			$force_refresh,
+		);
+
+		if ( ! $this->is_valid_cached_incentive( $incentive ) ) {
+			return null;
+		}
+
+		return $incentive;
+	}
+
+	/**
+	 * Fetches eligible connect incentive from the server.
+	 *
+	 * @return array|null Array of eligible incentive data or null.
+	 */
+	public function fetch_connect_incentive(): ?array {
+		// Request incentive from WCPAY API.
+		$url = add_query_arg(
+			[
+				// Store ISO-2 country code, e.g. `US`.
+				'country'      => WC()->countries->get_base_country(),
+				// Store locale, e.g. `en_US`.
+				'locale'       => get_locale(),
+				// WCPay active for duration in seconds.
+				'active_for'   => time() - get_option( 'wcpay_activation_timestamp', time() ),
+				// Whether the store has paid orders in the last 90 days.
+				'has_orders'   => ! empty(
+					wc_get_orders(
+						[
+							'status'       => [ 'wc-completed', 'wc-processing' ],
+							'date_created' => '>=' . strtotime( '-90 days' ),
+							'return'       => 'ids',
+							'limit'        => 1,
+						]
+					)
+				),
+				// Whether the store has at least one payment gateway enabled.
+				'has_payments' => ! empty( WC()->payment_gateways()->get_available_payment_gateways() ),
+			],
+			'https://public-api.wordpress.com/wpcom/v2/wcpay/incentives',
+		);
+
+		$response = wp_remote_get(
+			$url,
+			[
+				'user-agent' => 'WCPay/' . WCPAY_VERSION_NUMBER . '; ' . get_bloginfo( 'url' ),
```

Comment:
Excellent! This way, we can discern between WooCommerce and WCPay on the server.